solutions/CF Save the Nature.mdx

```diff
@@ -0,0 +1,214 @@
+
+## Explanation
+
+Binary search on the answer for the minimum number of tickets, which we call $t$, to sell to make the total ecological contribution to at least $k$. 
+
+Greedily, it is optimal to maximize the total revenue from selling these tickets since that will minimize the total number of tickets. If it is impossible to sell at least $k$ tickets, then output $-1$.
+
+Additionally, the problem indicates that only $x\%$ of the $a$-th ticket and $y\%$ of the $b$-th ticket could contribute to the total ecological contribution. We can precompute the multipliers $pct$ (namely: x, y, and x+y) over all $t$ tickets, then greedily match the largest multiplier with highest value ticket by sorting the multipliers $pct$ and tickets $p$ in decreasing order. Our total ecological contribution is $\sum\limits_{i=1}^{t} p_i \cdot pct_i$.
+
+
+**Time Complexity:** $\mathcal{O}(N\log^2 N)$
+
+<LanguageSection>
+
+<CPPSection>
+
+### C++ Implementation
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+
+int n; vector<ll> p;
+
+ll x, a, y, b, k;
+
+bool works(int sellTickets){
+	vector<ll> percentages(sellTickets);
+	
+	// Add x% to every a-th ticket.
+	for(int i = a - 1; i < sellTickets; i += a){
+		percentages[i] += x;
+	}
+	
+	// Add y% to every b-th ticket
+	for(int i = b - 1; i < sellTickets; i += b){
+		percentages[i] += y;
+	}
+	
+	sort(percentages.begin(), percentages.end(), greater<ll>());
+	
+	ll cur = 0;
+	for(int i = 0; i < sellTickets; i++){
+		cur += percentages[i] * p[i] / 100;
+	}
+	
+	return cur >= k;
+}
+
+void solve(){
+	cin >> n;
+	p.resize(n);
+	
+	for(int i = 0; i < n; i++){
+		cin >> p[i];
+	}
+	
+	sort(p.begin(), p.end(), greater<ll>());
+	
+	cin >> x >> a >> y >> b >> k;
+	
+	int l = 0, r = n, ans = -1;
+	
+	while(l <= r){
+		int mid = l + (r - l) / 2;
+		if(works(mid)){
+			ans = mid;
+			r = mid - 1;
+		} else {
+			l = mid + 1;
+		}
+	}
+	
+	cout << ans << endl;
+}
+
+int main(){
+	int q; cin >> q;
+	for(int i = 1; i <= q; i++){
+		solve();
+	}
+}
+```
+
+</CPPSection>
+
+<JavaSection>
+    
+### Java Implementation
+
+```java
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+import java.util.Comparator;
+
+public final class SaveNature {
+	public static void main(String[] args) throws IOException {
+		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
+		int queryNum = Integer.parseInt(read.readLine());
+		for (int q = 0; q < queryNum; q++) {
+			int ticketNum = Integer.parseInt(read.readLine());
+			long[] tickets = Arrays.stream(read.readLine().split(" ")).mapToLong(Long::parseLong).toArray();
+			Arrays.sort(tickets);
+
+			// first the percentage of the proceeds, then the frequency
+			int[] prog1 = Arrays.stream(read.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
+			int[] prog2 = Arrays.stream(read.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
```

Comment:
lines 114, 118, 119 are too long

Reply:
@SansPapyrus683 pls fix

Reply:
i mean how would i even make those shorter, java's just like that lmao

Reply:
idk, maybe split into two lines?

Reply:
yeah, but like how